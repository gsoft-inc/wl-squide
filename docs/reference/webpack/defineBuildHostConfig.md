---
order: 80
toc:
    depth: 2-3
---

# defineBuildHostConfig

Creates a webpack [configuration object](https://webpack.js.org/concepts/configuration/) that is adapted for a Squide host application in **build** mode.

!!!info
This function is a wrapper built on top of [@workleap/web-configs](https://www.npmjs.com/package/@workleap/webpack-configs). Make sure to read the [defineBuildConfig](https://gsoft-inc.github.io/wl-web-configs/webpack/configure-build/) documentation first.
!!!

## Reference

```ts
const webpackConfig = defineBuildHostConfig(swcConfig: {}, applicationName, options?: {})
```

## Parameters

- `swcConfig`: An SWC [configuration object](https://swc.rs/docs/configuration/swcrc).
- `applicationName`: The host application name.
- `options`: An optional object literal of options:
    - Accepts most of webpack `definedBuildConfig` [predefined options](https://gsoft-inc.github.io/wl-web-configs/webpack/configure-build/#3-set-predefined-options).
    - `htmlWebpackPluginOptions`: An optional object literal accepting any property of the [HtmlWebpackPlugin](https://github.com/jantimon/html-webpack-plugin#options).
    - `features`: An optional object literal of feature switches to define additional shared dependencies.
        - `i18next`: Whether or not to add `@squide/i18next` as a shared dependency.
    - `sharedDependencies`: An optional object literal of additional (or updated) module federation shared dependencies.
    - `moduleFederationPluginOptions`: An optional object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.

## Returns

A webpack [configuration object](https://webpack.js.org/concepts/configuration/) tailored for a Squide host application in build mode.

## Default shared dependencies

The `defineBuildHostConfig` function will add the following shared dependencies as `singleton` by default:
- [react](https://www.npmjs.com/package/react)
- [react-dom](https://www.npmjs.com/package/react-dom)
- [react-router-dom](https://www.npmjs.com/package/react-router-dom)
- [@squide/core](https://www.npmjs.com/package/@squide/core)
- [@squide/react-router](https://www.npmjs.com/package/@squide/react-router)
- [@squide/webpack-module-federation](https://www.npmjs.com/package/@squide/webpack-module-federation)
- [@squide/msw](https://www.npmjs.com/package/@squide/msw)

For the full shared dependencies configuration, have a look at the [defineConfig.ts](https://github.com/gsoft-inc/wl-squide/blob/main/packages/firefly/src/defineConfig.ts) file on GitHub.

## Usage

### Define a webpack config

```js !#6 host/webpack.build.js
// @ts-check

import { defineBuildHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.build.js";

export default defineBuildHostConfig(swcConfig, "host");
```

### Activate additional features

```js !#7-9 host/webpack.build.js
// @ts-check

import { defineBuildHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.build.js";

export default defineBuildHostConfig(swcConfig, "host", {
    features: {
        i18next: true
    }
});
```


!!!info
Features must be activated on the host application as well as every remote module.
!!!

### Specify additional shared dependencies

```js !#7-11 host/webpack.build.js
// @ts-check

import { defineBuildHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.build.js";

export default defineBuildHostConfig(swcConfig, "host", {
    sharedDependencies: {
        "@sample/shared": {
            singleton: true
        }
    }
});
```

!!!info
Additional shared dependencies must be configured on the host application as well as every remote module.
!!!

### Extend a default shared dependency

```js !#7-11 host/webpack.build.js
// @ts-check

import { defineBuildHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.build.js";

export default defineBuildHostConfig(swcConfig, "host", {
    sharedDependencies: {
        "react": {
            strictVersion: "18.2.0"
        }
    }
});
```

In the previous example, the `react` shared dependency will be **augmented** with the newly provided `strictVersion` option. The resulting shared dependency will be:

```js !#5
{
    "react": {
        eager: true,
        singleton: true,
        strictVersion: "18.2.0"
    }
}
```

### Override a default shared dependency

```js !#7-11 host/webpack.build.js
// @ts-check

import { defineBuildHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.build.js";

export default defineBuildHostConfig(swcConfig, "host", {
    sharedDependencies: {
        "react": {
            singleton: false
        }
    }
});
```

In the previous example, the `react` shared dependency `singleton` option will be **overrided** by the newly provided value. The resulting shared dependency will be:

```js !#4
{
    "react": {
        eager: true,
        singleton: false
    }
}
```

### Customize module federation configuration

While you could customize the [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) configuration by providing your own object literal through the `moduleFederationPluginOptions` option, we recommend using the `defineHostModuleFederationPluginOptions(applicationName, options)` function as it will take care of **merging** the custom options with the default plugin options.

```js !#7-9 host/webpack.build.js
// @ts-check

import { defineBuildHostConfig, defineHostModuleFederationPluginOptions } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.build.js";

export default defineBuildHostConfig(swcConfig, "host", {
    moduleFederationPluginOptions: defineHostModuleFederationPluginOptions("host", {
        runtime: "my-runtime-name"
    })
});
```

- `applicationName`: The host application name.
- `moduleFederationPluginOptions`: An object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.



