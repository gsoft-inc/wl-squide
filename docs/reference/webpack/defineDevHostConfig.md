---
order: 100
toc:
    depth: 2-3
---

# defineDevHostConfig

Creates a webpack [configuration object](https://webpack.js.org/concepts/configuration/) that is adapted for a Squide host application in **development** mode.

!!!info
This function is a wrapper built on top of [@workleap/web-configs](https://www.npmjs.com/package/@workleap/webpack-configs). Make sure to read the [defineDevConfig](https://gsoft-inc.github.io/wl-web-configs/webpack/configure-dev/) documentation first.
!!!

## Reference

```ts
const webpackConfig = defineDevHostConfig(swcConfig: {}, applicationName, port, options?: {})
```

## Parameters

- `swcConfig`: An SWC [configuration object](https://swc.rs/docs/configuration/swcrc).
- `applicationName`: The host application name.
- `port`: The host application port.
- `options`: An optional object literal of options:
    - Accepts most of webpack `definedDevConfig` [predefined options](https://gsoft-inc.github.io/wl-web-configs/webpack/configure-dev/#3-set-predefined-options).
    - `htmlWebpackPluginOptions`: An optional object literal accepting any property of the [HtmlWebpackPlugin](https://github.com/jantimon/html-webpack-plugin#options).
    - `features`: An optional object literal of feature switches to define additional shared dependencies.
        - `i18next`: Whether or not to add `@squide/i18next` as a shared dependency.
    - `sharedDependencies`: An optional object literal of additional (or updated) module federation shared dependencies.
    - `moduleFederationPluginOptions`: An optional object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.

## Returns

A webpack [configuration object](https://webpack.js.org/concepts/configuration/) tailored for a Squide host application in development mode.

## Default shared dependencies

The `defineDevHostConfig` function will add the following shared dependencies as `singleton` by default:
- [react](https://www.npmjs.com/package/react)
- [react-dom](https://www.npmjs.com/package/react-dom)
- [react-router-dom](https://www.npmjs.com/package/react-router-dom)
- [@squide/core](https://www.npmjs.com/package/@squide/core)
- [@squide/react-router](https://www.npmjs.com/package/@squide/react-router)
- [@squide/webpack-module-federation](https://www.npmjs.com/package/@squide/webpack-module-federation)
- [@squide/msw](https://www.npmjs.com/package/@squide/msw)

For the full shared dependencies configuration, have a look at the [defineConfig.ts](https://github.com/gsoft-inc/wl-squide/blob/main/packages/firefly/src/defineConfig.ts) file on Github.

## Usage

### Define a webpack config

```js !#6 host/webpack.dev.js
// @ts-check

import { defineDevHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevHostConfig(swcConfig, "host", 8080);
```

### Activate optional features

```js !#7-9 host/webpack.dev.js
// @ts-check

import { defineDevHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevHostConfig(swcConfig, "host", 8080, {
    features: {
        i18next: true
    }
});
```

!!!info
Features must be activated on the host application as well as every remote module.
!!!

### Specify additional shared dependencies

```js !#7-11 host/webpack.dev.js
// @ts-check

import { defineDevHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevHostConfig(swcConfig, "host", 8080, {
    sharedDependencies: {
        "@sample/shared": {
            singleton: true
        }
    }
});
```

!!!info
Additional shared dependencies must be configured on the host application as well as every remote module.
!!!

### Extend a default shared dependency

```js !#7-11 host/webpack.dev.js
// @ts-check

import { defineDevHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevHostConfig(swcConfig, "host", 8080, {
    sharedDependencies: {
        "react": {
            requiredVersion: "18.2.0"
        }
    }
});
```

In the previous example, the `react` shared dependency will be **augmented** with the newly provided `strictVersion` option. The resulting shared dependency will be:

```js !#5
{
    "react": {
        eager: true,
        singleton: true,
        requiredVersion: "18.2.0"
    }
}
```

### Override a default shared dependency

```js !#7-11 host/webpack.dev.js
// @ts-check

import { defineDevHostConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevHostConfig(swcConfig, "host", 8080, {
    sharedDependencies: {
        "react": {
            singleton: false
        }
    }
});
```

In the previous example, the `react` shared dependency `singleton` option will be **overrided** by the newly provided value. The resulting shared dependency will be:

```js !#4
{
    "react": {
        eager: true,
        singleton: false
    }
}
```

### Customize module federation configuration

While you could customize the [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) configuration by providing your own object literal through the `moduleFederationPluginOptions` option, we recommend using the `defineHostModuleFederationPluginOptions(applicationName, options)` function as it will take care of **merging** the custom options with the default plugin options.

```js !#7-9 host/webpack.dev.js
// @ts-check

import { defineDevHostConfig, defineHostModuleFederationPluginOptions } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevHostConfig(swcConfig, "host", 8080, {
    moduleFederationPluginOptions: defineHostModuleFederationPluginOptions("host", {
        runtime: "my-runtime-name"
    })
});
```

- `applicationName`: The host application name.
- `moduleFederationPluginOptions`: An object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.



