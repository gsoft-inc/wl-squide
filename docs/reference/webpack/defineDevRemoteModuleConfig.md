---
order: 90
toc:
    depth: 2-3
---

# defineDevRemoteModuleConfig

Creates a webpack [configuration object](https://webpack.js.org/concepts/configuration/) that is adapted for a Squide remote module application in **development** mode.

!!!warning
This function is a wrapper built on top of [@workleap/web-configs](https://www.npmjs.com/package/@workleap/webpack-configs). Make sure to read the [defineDevConfig](https://gsoft-inc.github.io/wl-web-configs/webpack/configure-dev/) documentation first.
!!!

## Reference

```ts
const webpackConfig = defineDevRemoteModuleConfig(swcConfig: {}, applicationName, port, options?: {})
```

## Parameters

- `swcConfig`: An SWC [configuration object](https://swc.rs/docs/configuration/swcrc).
- `applicationName`: The remote module application name.
- `port`: The remote module application port.
- `options`: An optional object literal of options:
    - Accepts most of webpack `definedDevConfig` [predefined options](https://gsoft-inc.github.io/wl-web-configs/webpack/configure-dev/#3-set-predefined-options).
    - `sharedDependencies`: An optional object literal of additional (or updated) module federation shared dependencies.
    - `moduleFederationPluginOptions`: An optional object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.

## Returns

A webpack [configuration object](https://webpack.js.org/concepts/configuration/) tailored for a Squide remote module application in development mode.

## Conventions

To fulfill Squide remote module requirements, the `defineDevRemoteModuleConfig` function will pre-configure the [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) with the following `filename` and `exposes` properties.

```js !#6-9
import ModuleFederationPlugin from "webpack/lib/container/ModuleFederationPlugin.js";

export default {
    plugins: [
        ModuleFederationPlugin({
            filename: "/remoteEntry.js",
            exposes: {
                "register.js": "./src/register"
            }
        })
    ]
};
```

!!!info
If the remote module `port` is `8081`, the remote module bundle is available at `http://localhost:8081/remoteEntry.js`.
!!!

## Default shared dependencies

The `defineDevRemoteModuleConfig` function will add the following shared dependencies as `singleton` by default:
- [react](https://www.npmjs.com/package/react)
- [react-dom](https://www.npmjs.com/package/react-dom)
- [react-router-dom](https://www.npmjs.com/package/react-router-dom)
- [@squide/core](https://www.npmjs.com/package/@squide/core)
- [@squide/react-router](https://www.npmjs.com/package/@squide/react-router)
- [@squide/webpack-module-federation](https://www.npmjs.com/package/@squide/webpack-module-federation)
- [@squide/msw](https://www.npmjs.com/package/@squide/msw)
- [@squide/firefly](https://www.npmjs.com/package/@squide/firefly)

For the full shared dependencies configuration, have a look at the [defineConfig.ts](https://github.com/gsoft-inc/wl-squide/blob/main/packages/webpack-module-federation/src/defineConfig.ts) file on Github.

## Usage

### Define a webpack config

```js !#6 remote-module/webpack.dev.js
// @ts-check

import { defineDevRemoteModuleConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevRemoteModuleConfig(swcConfig, "remote1", 8080);
```

### Specify additional shared dependencies

```js !#7-11 remote-module/webpack.dev.js
// @ts-check

import { defineDevRemoteModuleConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevRemoteModuleConfig(swcConfig, "remote1", 8080, {
    sharedDependencies: {
        "@sample/shared": {
            singleton: true
        }
    }
});
```

!!!info
Additional shared dependencies must be configured on the host application as well as every remote module.
!!!

### Extend a default shared dependency

```js !#7-11 remote-module/webpack.dev.js
// @ts-check

import { defineDevRemoteModuleConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevRemoteModuleConfig(swcConfig, "remote1", 8080, {
    sharedDependencies: {
        "react": {
            strictVersion: "18.2.0"
        }
    }
});
```

In the previous example, the `react` shared dependency will be **augmented** with the newly provided `strictVersion` option. The resulting shared dependency will be:

```js !#5
{
    "react": {
        eager: true,
        singleton: true,
        strictVersion: "18.2.0"
    }
}
```

### Override a default shared dependency

```js !#7-11 remote-module/webpack.dev.js
// @ts-check

import { defineDevRemoteModuleConfig } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevRemoteModuleConfig(swcConfig, "remote1", 8080, {
    sharedDependencies: {
        "react": {
            singleton: false
        }
    }
});
```

In the previous example, the `react` shared dependency `singleton` option will be **overrided** by the newly provided value. The resulting shared dependency will be:

```js !#4
{
    "react": {
        eager: true,
        singleton: false
    }
}
```

### Customize module federation configuration

While you could customize the [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) configuration by providing your own object literal through the `moduleFederationPluginOptions` option, we recommend using the `defineRemoteModuleFederationPluginOptions(applicationName, options)` function as it will take care of **merging** the custom options with the default plugin options.

```js !#7-9 remote-module/webpack.dev.js
// @ts-check

import { defineDevRemoteModuleConfig, defineRemoteModuleFederationPluginOptions } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevRemoteModuleConfig(swcConfig, "remote1", 8080, {
    moduleFederationPluginOptions: defineRemoteModuleFederationPluginOptions("remote1", {
        runtime: "my-runtime-name"
    })
});
```

- `applicationName`: The host application name.
- `moduleFederationPluginOptions`: An object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.

### Expose an additional module

```js !#7-11 remote-module/webpack.dev.js
// @ts-check

import { defineDevRemoteModuleConfig, defineRemoteModuleFederationPluginOptions } from "@squide/firefly/defineConfig.js";
import { swcConfig } from "./swc.dev.js";

export default defineDevRemoteModuleConfig(swcConfig, "remote1", 8080, {
        moduleFederationPluginOptions: defineRemoteModuleFederationPluginOptions("remote1", {
            exposes: {
                "./foo": "./src/bar"
            }
        })
});
```
