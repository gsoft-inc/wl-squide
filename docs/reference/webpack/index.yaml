order: -10
label: "webpack"
