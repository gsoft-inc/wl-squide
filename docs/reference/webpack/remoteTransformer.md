# remoteTransformer

## Reference

### Parameters

### Returns

## Usage
