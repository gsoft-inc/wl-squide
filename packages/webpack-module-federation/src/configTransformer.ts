import { RemoteEntryPoint, RemoteModuleName } from "./remoteDefinition.ts";

import { isNil } from "@squide/core";
import merge from "deepmerge";
import webpack from "webpack";

// Webpack doesn't export ModuleFederationPlugin typings.
export type ModuleFederationPluginOptions = ConstructorParameters<typeof webpack.container.ModuleFederationPlugin>[0];
export type SharedDependency = ModuleFederationPluginOptions["shared"];

export interface ModuleFederationOptions {
    router?: "react-router";
    pluginOptions?: ModuleFederationPluginOptions;
}

const DefaultSharedDependencies: SharedDependency = {
    "react": {
        singleton: true,
        eager: true
    },
    "react-dom": {
        singleton: true,
        eager: true
    },
    "@squide/core": {
        singleton: true,
        eager: true
    },
    "@squide/webpack-module-federation": {
        singleton: true,
        eager: true
    }
};

const ReactRouterSharedDependencies: SharedDependency = {
    "react-router-dom": {
        singleton: true,
        eager: true
    },
    "@squide/react-router": {
        singleton: true,
        eager: true
    }
};

function createSharedObject({ router = "react-router", pluginOptions = {} }: ModuleFederationOptions) {
    return merge.all<SharedDependency>([
        DefaultSharedDependencies,
        router === "react-router" ? ReactRouterSharedDependencies : {},
        pluginOptions.shared ?? {}
    ]);
}

function validateConfig(config: webpack.Configuration) {
    if (!isNil(config.plugins) && config.plugins.some(x => x.constructor.name === webpack.container.ModuleFederationPlugin.name)) {
        throw new Error("ModuleFederationPlugin has already been configured. Please remove ModuleFederationPlugin from your configuration plugins before calling this function.");
    }
}

export function hostTransformer(config: webpack.Configuration, name: string, options: ModuleFederationOptions = {}): webpack.Configuration {
    validateConfig(config);

    const pluginOptions: ModuleFederationPluginOptions = {
        name,
        shared: createSharedObject(options)
    };

    return {
        ...config,
        plugins: [
            ...(config.plugins ?? []),
            new webpack.container.ModuleFederationPlugin(pluginOptions)
        ]
    };
}

export function remoteTransformer(config: webpack.Configuration, name: string, options: ModuleFederationOptions = {}): webpack.Configuration {
    validateConfig(config);

    const pluginOptions: ModuleFederationPluginOptions = {
        name,
        filename: RemoteEntryPoint,
        exposes: {
            [RemoteModuleName]: "./src/register"
        },
        shared: createSharedObject(options)
    };

    return {
        ...config,
        plugins: [
            ...(config.plugins ?? []),
            new webpack.container.ModuleFederationPlugin(pluginOptions)
        ]
    };
}
