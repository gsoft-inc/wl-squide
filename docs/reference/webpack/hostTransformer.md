# hostTransformer

Add a [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) plugin instance configured with `@squide` conventions to a Webpack [configuration object](https://webpack.js.org/concepts/configuration/).

## Reference

```ts
hostTransformer(config, name, options?: { router?, pluginOptions? })
```

### Parameters

- `config`: A Webpack [configuration object](https://webpack.js.org/concepts/configuration/).
- `name`: The host application name.
- `options`: An optional object literal of options.
    - `router`: An optional router identifier (default to `"react-router"`).
    - `pluginOptions`: An optional object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.

### Returns

A transformed Webpack [configuration object](https://webpack.js.org/concepts/configuration/) bonified with a [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) plugin instance.

## Default shared dependencies

The `hostTransformer` function will add the following shared dependencies as `singleton` by default:
- [react](https://www.npmjs.com/package/react)
- [react-dom](https://www.npmjs.com/package/react-dom)
- [react-router-dom](https://www.npmjs.com/package/react-router-dom)
- [@squide/core](https://www.npmjs.com/package/@squide/core)
- [@squide/react-router](https://www.npmjs.com/package/@squide/react-router)
- [@squide/webpack-module-federation](https://www.npmjs.com/package/@squide/webpack-module-federation)

For their detailed configuration, have a look at the [configTransformer.ts](https://github.com/workleap/wl-squide/blob/main/packages/webpack-module-federation/src/configTransformer.ts){target="_blank"} file on Github.

## Usage

### Transform a Webpack config

```js !#9 webpack.config.js
import { hostTransformer } from "@squide/webpack-module-federation/configTransformer.js";

/** @type {import("webpack").Configuration} */
const config = {
    target: "web",
    entry: "./src/index.ts",
};

const federatedConfig = hostTransformer(config, "host");

export default federatedConfig;
```

### Specify additional shared dependencies

```js !#10-16 webpack.config.js
import { hostTransformer } from "@squide/webpack-module-federation/configTransformer.js";

/** @type {import("webpack").Configuration} */
const config = {
    target: "web",
    entry: "./src/index.ts",
};

const federatedConfig = hostTransformer(config, "host", {
    pluginOptions: {
        shared: {
            "@sample/shared": {
                singleton: true
            }
        }
    }
});

export default federatedConfig;
```

### Override a pre-configured shared dependency

```js !#12-14 webpack.config.js
import { hostTransformer } from "@squide/webpack-module-federation/configTransformer.js";

/** @type {import("webpack").Configuration} */
const config = {
    target: "web",
    entry: "./src/index.ts",
};

const federatedConfig = hostTransformer(config, "host", {
    pluginOptions: {
        shared: {
            "react": {
                strictVersion: "18.2.0"
            }
        }
    }
});

export default federatedConfig;
```
