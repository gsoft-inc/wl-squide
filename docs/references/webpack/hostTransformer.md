# hostTransformer

Add to an existing Webpack [configuration object](https://webpack.js.org/concepts/configuration/) a [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) plugin instance pre-configured for a `@squide` host application.

## Reference

```ts
const federatedConfig = hostTransformer(config, name, options?: { router?, pluginOptions? })
```

### Parameters

- `config`: A Webpack [configuration object](https://webpack.js.org/concepts/configuration/).
- `name`: The host application name.
- `options`: An optional object literal of options.
    - `router`: An optional router identifier (default is `"react-router"`).
    - `pluginOptions`: An optional object literal of [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) options.

### Returns

A Webpack [configuration object](https://webpack.js.org/concepts/configuration/) enriched with a [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/) plugin instance.

## Default shared dependencies

The `hostTransformer` function will add the following shared dependencies as `singleton` by default:
- [react](https://www.npmjs.com/package/react)
- [react-dom](https://www.npmjs.com/package/react-dom)
- [react-router-dom](https://www.npmjs.com/package/react-router-dom)
- [@squide/core](https://www.npmjs.com/package/@squide/core)
- [@squide/react-router](https://www.npmjs.com/package/@squide/react-router)
- [@squide/webpack-module-federation](https://www.npmjs.com/package/@squide/webpack-module-federation)

For the full shared dependencies configuration, have a look at the [configTransformer.ts](https://github.com/workleap/wl-squide/blob/main/packages/webpack-module-federation/src/configTransformer.ts) file on Github.

## Usage

### Transform a Webpack config

```js !#10 host/webpack.config.js
import { hostTransformer } from "@squide/webpack-module-federation/configTransformer.js";

/** @type {import("webpack").Configuration} */
const webpackConfig = {
    target: "web",
    entry: "./src/index.ts",
    ...
};

const federatedConfig = hostTransformer(config, "host");

export default federatedConfig;
```

### Specify additional shared dependencies

!!!info
Additional shared dependencies must be configured on the host application as well as every remote module.
!!!

```js !#11-17 host/webpack.config.js
import { hostTransformer } from "@squide/webpack-module-federation/configTransformer.js";

/** @type {import("webpack").Configuration} */
const config = {
    target: "web",
    entry: "./src/index.ts",
    ...
};

const federatedConfig = hostTransformer(config, "host", {
    pluginOptions: {
        shared: {
            "@sample/shared": {
                singleton: true
            }
        }
    }
});

export default federatedConfig;
```

### Override a default shared dependency

```js !#13-15 host/webpack.config.js
import { hostTransformer } from "@squide/webpack-module-federation/configTransformer.js";

/** @type {import("webpack").Configuration} */
const config = {
    target: "web",
    entry: "./src/index.ts",
    ...
};

const federatedConfig = hostTransformer(config, "host", {
    pluginOptions: {
        shared: {
            "react": {
                strictVersion: "18.2.0"
            }
        }
    }
});

export default federatedConfig;
```
