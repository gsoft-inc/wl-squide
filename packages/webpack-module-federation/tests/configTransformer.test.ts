import { hostTransformer, remoteTransformer } from "../src/configTransformer.ts";

import webpack from "webpack";

/*
TODO: Add snapshot tests with at least those use cases:
    - pluginOptions already contains a "shared" section
    - pluginOptions doesn't contains a "shared" section
    - with additional shared dependencies
    - with reactOptions (and all the other similar options)
    - with additional "exposes" for a remote
    - with a different filename for a remote

    - with pluginOptions that are not "shared" values
*/

class DummyPlugin {
    _options: unknown;

    constructor(options: unknown) {
        this._options = options;
    }

    apply() {
        console.log("Apply Dummy plugin.");
    }
}

describe("host", () => {
    test("default configuration", () => {
        expect(hostTransformer({}, "host")).toMatchSnapshot();
    });

    test("merge with existing config values", () => {
        const userConfig: webpack.Configuration = {
            mode: "development",
            target: "web",
            devtool: "eval-cheap-module-source-map"
        };

        expect(hostTransformer(userConfig, "host")).toMatchSnapshot();
    });

    test("when the plugins array already exist, append to the existing plugins array", () => {
        const userConfig: webpack.Configuration = {
            plugins: [new DummyPlugin({})]
        };

        expect(hostTransformer(userConfig, "host")).toMatchSnapshot();
    });

    test("when a ModuleFederationPlugin instance has already been added to the plugins, throw an error", () => {
        const userConfig: webpack.Configuration = {
            plugins: [
                new webpack.container.ModuleFederationPlugin({})
            ]
        };

        expect(() => hostTransformer(userConfig, "host")).toThrow(/ModuleFederationPlugin/);
    });
});

describe("remote", () => {
    test("default configuration", () => {
        expect(remoteTransformer({}, "remote")).toMatchSnapshot();
    });
});


