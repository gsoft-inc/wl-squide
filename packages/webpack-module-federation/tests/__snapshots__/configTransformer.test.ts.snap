// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`host default configuration 1`] = `
{
  "plugins": [
    ModuleFederationPlugin {
      "_options": {
        "name": "host",
        "shared": {
          "@squide/core": {
            "eager": true,
            "singleton": true,
          },
          "@squide/react-router": {
            "eager": true,
            "singleton": true,
          },
          "@squide/webpack-module-federation": {
            "eager": true,
            "singleton": true,
          },
          "react": {
            "eager": true,
            "singleton": true,
          },
          "react-dom": {
            "eager": true,
            "singleton": true,
          },
          "react-router-dom": {
            "eager": true,
            "singleton": true,
          },
        },
      },
    },
  ],
}
`;

exports[`host merge with existing config values 1`] = `
{
  "devtool": "eval-cheap-module-source-map",
  "mode": "development",
  "plugins": [
    ModuleFederationPlugin {
      "_options": {
        "name": "remote",
        "shared": {
          "@squide/core": {
            "eager": true,
            "singleton": true,
          },
          "@squide/react-router": {
            "eager": true,
            "singleton": true,
          },
          "@squide/webpack-module-federation": {
            "eager": true,
            "singleton": true,
          },
          "react": {
            "eager": true,
            "singleton": true,
          },
          "react-dom": {
            "eager": true,
            "singleton": true,
          },
          "react-router-dom": {
            "eager": true,
            "singleton": true,
          },
        },
      },
    },
  ],
  "target": "web",
}
`;

exports[`host when the plugins array already exist, append to the existing plugins 1`] = `
{
  "plugins": [
    DummyPlugin {
      "_options": {},
    },
    ModuleFederationPlugin {
      "_options": {
        "name": "remote",
        "shared": {
          "@squide/core": {
            "eager": true,
            "singleton": true,
          },
          "@squide/react-router": {
            "eager": true,
            "singleton": true,
          },
          "@squide/webpack-module-federation": {
            "eager": true,
            "singleton": true,
          },
          "react": {
            "eager": true,
            "singleton": true,
          },
          "react-dom": {
            "eager": true,
            "singleton": true,
          },
          "react-router-dom": {
            "eager": true,
            "singleton": true,
          },
        },
      },
    },
  ],
}
`;

exports[`remote default configuration 1`] = `
{
  "plugins": [
    ModuleFederationPlugin {
      "_options": {
        "exposes": {
          "./register": "./src/register",
        },
        "filename": "remoteEntry.js",
        "name": "remote",
        "shared": {
          "@squide/core": {
            "eager": true,
            "singleton": true,
          },
          "@squide/react-router": {
            "eager": true,
            "singleton": true,
          },
          "@squide/webpack-module-federation": {
            "eager": true,
            "singleton": true,
          },
          "react": {
            "eager": true,
            "singleton": true,
          },
          "react-dom": {
            "eager": true,
            "singleton": true,
          },
          "react-router-dom": {
            "eager": true,
            "singleton": true,
          },
        },
      },
    },
  ],
}
`;
